### ISSUE TITLE
IAN library and OpenClipart symbols can no longer be inserted.

### ISSUE THREAD

**Issue by adrianbj** [SVG-Edit/svgedit#274]
It worked fine in 2.8.1, but testing with the 3.x branch, it no longer works.

Interestingly, the smiley and logo demo images still work fine. Is there something that needs to be changed in SVGEdit, or in the external libraries?

Thanks!

**Comment by adrianbj** [comment]
I have solved this for the IAN Library to adjusting the `postMessage` code included in the IAN's iframe interface. Not sure who needs to take care of things for OpenClipart, so I'll leave this open for now.

The other problem here is that any pre v3.x installations will no longer be able to access the IAN libraries now that I have made those changes. I wonder if it might be possible to support both somehow?

**Comment by brettz9** [comment]
The smiley and logo images are local copies so that's why they should be working regardless.

Apologies also for the breaking code. It looks like the API change that could cause a problem is the fact that the old code rejects with the presence of a `namespace` property and the new code requires it. If you recall other changes, please let me know. We might solve this by having the new approach send a differently named property like "namespace-key" or something.

Btw, I am not able to successfully login on Chrome (within svgedit), despite having clicked the activation email and being able to login outside of the app and within Firefox; I don't know if the issue is on our end, if iframe inclusion is blocked on your end, but it only works to get past the login screen in Firefox for me.

I'd be very grateful for a reply, as I'd really like to get this working before we finalize our 3.0.0 release...

**Comment by adrianbj** [comment]
Hi @brettz9 - sorry I didn't respond sooner.

Yes, I think the namespace was probably the only change I needed to make.

I just tested logging into the IAN libraries via Chrome and Firefox and both worked here, so not sure what might be going wrong at your end.

Please have a play around with inserting a few symbols though if you can - I have actually noticed several other bugs introduced since 2.5 that affect symbols. Some of these bugs are browser specific, like the bounding box shift issue which is actually now fixed on Chrome in 3.x, but it still an issue in Firefox.

**Comment by adrianbj** [comment]
Hey @brettz9 - does this mean I should remove the namespace from my script? Will let let it work with both 2.5 and 3.x versions of SVGEdit?

**Comment by brettz9** [comment]
Firstly, I just pushed a fix for my code that closed this issue.

In reply to your question, yes, you can now remove `namespace` from your script, but to work with the current code as well, you will need to add a `"namespace-key": "imagelib"`. This should give you compatibility with both versions.

A few other things...

1. In order to address a potential security issues in Imagelib (#288), I've just prevented  custom entity declarations (as may be added within an svg DOCTYPE). If you find that a number of images actually use this XML (but not HTML) capability, let me know, and I may be able to come up with a less heavy-handed prevention (by not stripping out all custom entity declarations).
2. For compatibility with IE9, we're still posting (and accepting) strings with `postMessage`. Since I expect we may only post messages as objects in our own code in the future, I made a forward-compatible change to accept messages posted by object even though we're not doing that now internally. But you can continue to post messages with strings.
3. As far as the symbol breakage, please feel free to continue to report any other issues you encounter. Unfortunately, my energy and expertise are not focused on much of the lower level functions that may be responsible, though if someone can pinpoint the version and code responsible for breakage, I hope to be able to help fix things without losing whatever good changes may have been intended by the breaking code.

**Comment by brettz9** [comment]
And thanks for drawing our attention to this issue! It's valuable to have some compatibility again, even if I can't get it logging in on Chrome...

**Comment by adrianbj** [comment]
> even if I can't get it logging in on Chrome...

Do you think there is something I can do at my end?

Are you getting any console errors? I feel like it must be session/cookie related - maybe something CORS. Does it help if you use the Chrome plugin: https://chrome.google.com/webstore/detail/allow-control-allow-origi/nlfbmbojpeacfghkpbjhddihlkkiljbi?hl=en

**Comment by brettz9** [comment]
I see the old API appears to need a string; I'll see about addressing. I think I will get things working with the old API, but provide a parameter to you from the new code which you could use to send the new object-based API (with `namespace` too), so that for any next version, you will already be compatible.

The CORS plugin makes no difference, and I am not getting console errors.

When I first visit select IAN, I get the URL:

> https://ian.umces.edu/symbols/catalog/svgedit/album_chooser.php

Here are the request headers (I added a cache buster "dddd" to the URL in this test, but all steps are the same, with or without my adding this cache buster):

```
GET /symbols/catalog/svgedit/album_chooser.php?dddd=1537837572020 HTTP/1.1
Host: ian.umces.edu
Connection: keep-alive
Pragma: no-cache
Cache-Control: no-cache
Upgrade-Insecure-Requests: 1
User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_13_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/69.0.3497.100 Safari/537.36
Sec-Metadata: cause="user-activated", destination="document", target="nested", site="cross-site"
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8
Referer: http://localhost:8000/editor/svg-editor-es.html
Accept-Encoding: gzip, deflate, br
Accept-Language: en-US,en;q=0.9
```

Here are the response headers:

```
HTTP/1.1 200 OK
Date: Tue, 25 Sep 2018 01:06:13 GMT
Server: Apache/2.2.22
Set-Cookie: PHPSESSID=34892b571c821255b73556b9ef11d168; path=/
Expires: Thu, 19 Nov 1981 08:52:00 GMT
Cache-Control: no-store, no-cache, must-revalidate, post-check=0, pre-check=0
Pragma: no-cache
Set-Cookie: ce6c119e2ab868610fb91fcad053c97e=1c97fa0a61923d9ece26cd19dde6fddc; expires=Tue, 09-Oct-2018 01:06:13 GMT; path=/
Vary: Accept-Encoding
Content-Encoding: gzip
Content-Length: 2592
Keep-Alive: timeout=15, max=100
Connection: Keep-Alive
Content-Type: text/html; charset=utf-8
```

(I see `expires` here is for 1981; maybe this should be changed? e.g., as per https://stackoverflow.com/questions/4574334/change-the-expiry-time-of-phpsessid-cookie )

At such a time I am presented with login credentials (and the Network tab of the console confirms that the preview of this page indeed includes the login form):

However, when I visit this page directly in a separate tab, it is logged in and properly showing the category choices so the system knows I'm logged in, but for some reason, within the editor, the same URL (https://ian.umces.edu/symbols/catalog/svgedit/album_chooser.php) gives (and keeps giving) the login page.

When visiting in its own page, here are the headers (Note it sends with a cookie header in this context):

Request:

```
GET /symbols/catalog/svgedit/album_chooser.php HTTP/1.1
Host: ian.umces.edu
Connection: keep-alive
Pragma: no-cache
Cache-Control: no-cache
Upgrade-Insecure-Requests: 1
User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_13_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/69.0.3497.100 Safari/537.36
Sec-Metadata: cause="forced", destination="document", target="top-level", site="same-origin"
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8
Accept-Encoding: gzip, deflate, br
Accept-Language: en-US,en;q=0.9
Cookie: __utmc=133193622; __utmz=133193622.1536835882.1.1.utmcsr=(direct)|utmccn=(direct)|utmcmd=(none); PHPSESSID=fa1fbd98b90b528924834c97421e95ac; 5485d2542691066cce9588c55e8730f8=b376d64fe66794a9b6b164accf0c8e30; ce6c119e2ab868610fb91fcad053c97e=960e8494f2006059a9c51f8aa1931fef; iilcpg_data=YTozOntzOjI6IklEIjtzOjMyOiI4YTM2MTNkNDZjY2E5NzcyMTAwYTA2NzI5YTUzZThjZSI7czoyOiJhbSI7aToxO3M6NDoibGFuZyI7czo3OiJlbmdsaXNoIjt9; __utma=133193622.85920508.1536835882.1537817114.1537835706.3; __utmt=1; __utmb=133193622.6.10.1537835706
```

Response:

```
HTTP/1.1 200 OK
Date: Tue, 25 Sep 2018 01:25:54 GMT
Server: Apache/2.2.22
Expires: Thu, 19 Nov 1981 08:52:00 GMT
Cache-Control: no-store, no-cache, must-revalidate, post-check=0, pre-check=0
Pragma: no-cache
Vary: Accept-Encoding
Content-Encoding: gzip
Content-Length: 4376
Keep-Alive: timeout=15, max=98
Connection: Keep-Alive
Content-Type: text/html; charset=utf-8
```

Ok, back to the problematic SVG-Edit request/responses...After submitting, the form, I get the following in the Network tab of console:

1) The following which shows "Failed to load response data" in its preview:

> https://ian.umces.edu/imagelibrary/login.php?referer=/symbols/catalog/svgedit/album_chooser.php

Here are its request headers:

```
POST /imagelibrary/login.php?referer=/symbols/catalog/svgedit/album_chooser.php HTTP/1.1
Host: ian.umces.edu
Connection: keep-alive
Content-Length: 66
Pragma: no-cache
Cache-Control: no-cache
Origin: https://ian.umces.edu
Upgrade-Insecure-Requests: 1
Content-Type: application/x-www-form-urlencoded
User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_13_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/69.0.3497.100 Safari/537.36
Sec-Metadata: cause="user-activated", destination="document", target="nested", site="same-origin"
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8
Referer: https://ian.umces.edu/symbols/catalog/svgedit/album_chooser.php?dddd=1537837572020
Accept-Encoding: gzip, deflate, br
Accept-Language: en-US,en;q=0.9
```

Here are its response headers:

```
HTTP/1.1 200 OK
Date: Tue, 25 Sep 2018 01:06:36 GMT
Server: Apache/2.2.22
Set-Cookie: ce6c119e2ab868610fb91fcad053c97e=e87c169791e8128c78d9137683e0c4e9; expires=Tue, 09-Oct-2018 01:06:37 GMT; path=/
Vary: Accept-Encoding
Content-Encoding: gzip
Content-Length: 102
Keep-Alive: timeout=15, max=100
Connection: Keep-Alive
Content-Type: text/html; charset=utf-8
```

2) And then the following shows up after the above, still showing the login details in the response preview:
> https://ian.umces.edu/symbols/catalog/svgedit/album_chooser.php

Here are its request headers:

```
GET /symbols/catalog/svgedit/album_chooser.php HTTP/1.1
Host: ian.umces.edu
Connection: keep-alive
Pragma: no-cache
Cache-Control: no-cache
Upgrade-Insecure-Requests: 1
User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_13_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/69.0.3497.100 Safari/537.36
Sec-Metadata: cause="forced", destination="document", target="nested", site="same-origin"
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8
Referer: https://ian.umces.edu/imagelibrary/login.php?referer=/symbols/catalog/svgedit/album_chooser.php
Accept-Encoding: gzip, deflate, br
Accept-Language: en-US,en;q=0.9
```

And its response headers:

```
HTTP/1.1 200 OK
Date: Tue, 25 Sep 2018 01:06:37 GMT
Server: Apache/2.2.22
Set-Cookie: PHPSESSID=8cd14c18802b95c7e85cb8e95201813f; path=/
Expires: Thu, 19 Nov 1981 08:52:00 GMT
Cache-Control: no-store, no-cache, must-revalidate, post-check=0, pre-check=0
Pragma: no-cache
Set-Cookie: ce6c119e2ab868610fb91fcad053c97e=31b69563a139e4dc412f17995ca4ee77; expires=Tue, 09-Oct-2018 01:06:37 GMT; path=/
Vary: Accept-Encoding
Content-Encoding: gzip
Content-Length: 2592
Keep-Alive: timeout=15, max=99
Connection: Keep-Alive
Content-Type: text/html; charset=utf-8
```

Could the issue be with redirecting because when I remove the following in the Chrome Inspector:

> ?referer=/symbols/catalog/svgedit/album_chooser.php

...and then submit my credentials, it logs in to a different page properly.

If that is not it and the problem is specific to `album_chooser.php`, I see your page uses `document.write` and I know that can cause page refreshes---could that code be changed?

Hopefully that may help with debugging--and again, I'll try to get this genuinely working backward-compatibly...

**Comment by brettz9** [comment]
I'll reopen until we confirm you can change the API on your side...

**Comment by brettz9** [comment]
Btw, Openclipart https://openclipart.org/developers does have a developers page, so we could reimplement with our own interface....

**Comment by brettz9** [comment]
Sorry, I realized I had one other issue with the old API--expect it should be fixed now...

**Comment by adrianbj** [comment]
Hi @brettz9 - sorry for the delay in getting back to you.

I have a solution that is working for both:
https://rawgit.com/SVG-Edit/svgedit/master/editor/svg-editor.html and https://svg-edit.github.io/svgedit/releases/svg-edit-2.8.1/svg-editor.html

Unfortunately it's a little ugly. I don't have time right now to go through and figure out what's what, so I went with two completely separate sets of logic based on your `svgedit=3` 

```
if(svgEditVerson == 3) {

	(function () {
	  'use strict';

	  var _extends = Object.assign || function (target) {
	    for (var i = 1; i < arguments.length; i++) {
	      var source = arguments[i];

	      for (var key in source) {
	        if (Object.prototype.hasOwnProperty.call(source, key)) {
	          target[key] = source[key];
	        }
	      }
	    }

	    return target;
	  };

	  /* globals jQuery */
	  var $ = jQuery;
	  $('a').live('click', function() {
	    var href = this.href;

	    var target = window.top;

	    var post = function post(message) {
			// Todo: Make origin customizable as set by opening window
			// Todo: If dropping IE9, avoid stringifying
			target.postMessage(_extends({
				namespace: 'imagelib'
			}, message), '*');

	    };
	    // Convert Non-SVG images to data URL first
	    // (this could also have been done server-side by the library)
	    // Send metadata (also indicates file is about to be sent)
	    post({
	      name: $(this).text(),
	      id: href,
		  preview_url: href.replace("normal_","thumb_")
	    });
	    if (!this.href.includes('.svg')) {
	      var img = new Image();
	      img.onload = function () {
	        var canvas = document.createElement('canvas');
	        canvas.width = this.width;
	        canvas.height = this.height;
	        // load the raster image into the canvas
	        canvas.getContext('2d').drawImage(this, 0, 0);
	        // retrieve the data: URL
	        var data = void 0;
	        try {
	          data = canvas.toDataURL();
	        } catch (err) {
	          // This fails in Firefox with file:// URLs :(
	          alert('Data URL conversion failed: ' + err);
	          data = '';
	        }
	        post({ href: href, data: data });
	      };
	      img.src = href;
	    } else {
	      // Do ajax request for image's href value
	      $.get(href, function (data) {
	        post({ href: href, data: data });
	      }, 'html'); // 'html' is necessary to keep returned data as a string
	    }
	    return false;
	  });

	}());

}
else {

	$('a').live('click', function() {

		var href = this.href;

		if(href.indexOf('http') !== -1) {

		var target = window.parent.parent;
		// Convert Non-SVG images to data URL first
		// (this could also have been done server-side by the library)
		if(this.href.indexOf('.svg') === -1) {

			var meta_str = JSON.stringify({
				name: $(this).text(),
				id: href,
				preview_url: href.replace("normal_","thumb_")
			});
			//window.top.postMessage(meta_str, "*");
			target.postMessage(meta_str, "*");

			var img = new Image();
			img.onload = function() {
				var canvas = document.createElement("canvas");
				canvas.width = this.width;
				canvas.height = this.height;
				// load the raster image into the canvas
				canvas.getContext("2d").drawImage(this,0,0);
				// retrieve the data: URL
				try {
					var dataurl = canvas.toDataURL();
				} catch(err) {
					// This fails in Firefox with file:// URLs :(
					alert("Data URL conversion failed: " + err);
					var dataurl = "";
				}
				//window.top.postMessage('|' + href + '|' + dataurl, "*");
				target.postMessage('|' + href + '|' + dataurl, "*");
			}
			img.src = href;
		} else {
			// Send metadata (also indicates file is about to be sent)
			var meta_str = JSON.stringify({
				name: $(this).text(),
				id: href,
				preview_url: dirname(href)+"/thumb_"+basename(href,'svg')+"png"
			});
			//window.top.postMessage(meta_str, "*");
			target.postMessage(meta_str, "*");
			// Do ajax request for image's href value
			$.get(href, function(data) {
				data = '|' + href + '|' + data;
				// This is where the magic happens!
				//window.top.postMessage(data, "*");
				target.postMessage(data, "*");

			}, 'html'); // 'html' is necessary to keep returned data as a string
		}
		return false;

		}

	});

}
```

**Comment by brettz9** [comment]
Nice work! Awesome to hear it is working again for both SVG-Edit versions... And let me know if there is anything further I can do to help debugging my Chrome issue...

And FWIW, I've started some code to allow Openclipart to work again using their API...

Btw, consolidating opportunities aside, there does seem to be different logic between your code for version 3 and prior as far as how SVG vs. non-SVG image paths are determined, while I would have expected those to be the same regardless of version. The images I chose worked in my testing, but I wonder if this is a sign there could be problems say when a non-SVG image is chosen or what not...

**Comment by brettz9** [comment]
I still see a warning about `document.write` though it appears that may be because of some internal Chrome code, despite originating from your site:

> VM437 jsapi:23 [Violation] Avoid using document.write().
> google.loader.f @ VM437 jsapi:23
> O.load @ VM437 jsapi:37
> google.load @ VM437 jsapi:18
> (anonymous) @ album_chooser.php?svgedit=3:28

FYI, there is also this warning:
> album_chooser.php?svgedit=3:1 [DOM] Input elements should have autocomplete attributes (suggested: "current-password"): (More info: https://goo.gl/9p2vKq) <input type=​"password" name=​"password" size=​"30">​

I'd think the session issues may be more likely though...

**Comment by adrianbj** [comment]
So weird - I am not seeing either of these. Where are these showing up for you?

I'll get onto the session issue in about half an hour - need to leave my desk for a bit.

**Comment by brettz9** [comment]
I'm also on MacOS, though my Chrome version is 69.0.3497.100 (and says it's up to date)...

![screen shot 2018-09-26 at 9 47 04 am](https://user-images.githubusercontent.com/20234/46052570-9cbe9100-c171-11e8-931e-4d5a02aefc7e.png)

If I disable my VPN (from China), I get the following errors now instead (probably because Google is blocked here):

> GET https://www.google.com/jsapi net::ERR_CONNECTION_TIMED_OUT
> ian.umces.edu/symbols/catalog/svgedit/album_chooser.php?svgedit=3:28 Uncaught ReferenceError: google is not defined
>     at ian.umces.edu/symbols/catalog/svgedit/album_chooser.php?svgedit=3:28
> (anonymous) @ ian.umces.edu/symbols/catalog/svgedit/album_chooser.php?svgedit=3:28
> ian.umces.edu/symbols/catalog/svgedit/album_chooser.php?svgedit=3:1 [DOM] Input elements > should have autocomplete attributes (suggested: "current-password"): (More info: https://goo.gl/9p2vKq) input

**Comment by adrianbj** [comment]
Thanks for the screenshot - I didn't have LogXMLHttpRequests checked. Now I see those errors. I changed the way jquery is loaded so it is no longer triggering the document.write warning.

I am wondering if you could test something else for me please. If you login at: http://ian.umces.edu/imagelibrary/ and check "Remember me" and then try to access the libraries via SVGEdit does it let you in?

**Comment by brettz9** [comment]
Ok, firstly, I'm no longer getting the `document.write` errors, so nice to have that excluded.

Secondly, yes, I've already been logged in at the separate page, but for some reason, when it appears in that frame (within svgedit), it acts as though it is not logged in, the first time as well as the second time.

And when I log out at the independent page and try things fresh in svgedit, it doesn't work that way either.

It's like Chrome is treating the frame as credential-less... And that is backed up by the fact that the "Cookie:" header only works for me outside of svgedit...

**Comment by adrianbj** [comment]
This is a longshot, but could you try now please?

You will see that there is a document.write back again, but it's related to an older version of google analytics that I don't want to mess with at the moment and since we know this is not the issue, I decided to reinstate.

**Comment by adrianbj** [comment]
Also wondering if maybe you have some custom security settings enabled on your Chrome setup? Would you be willing to reset your browser settings?

Also, I wonder if it makes any difference if you are in an Incognito window?

**Comment by adrianbj** [comment]
Thanks for trying.

I just made some Access-Control-Allow-Origin changes. Would you mind trying again please?

If that doesn't help, I'll look into the Sec-Metdata thing.

**Comment by brettz9** [comment]
Now I'm getting an Internal Security Error on first load:

Request:

```
GET /symbols/catalog/svgedit/album_chooser.php?svgedit=3 HTTP/1.1
Host: ian.umces.edu
Connection: keep-alive
Pragma: no-cache
Cache-Control: no-cache
Upgrade-Insecure-Requests: 1
User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_13_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/69.0.3497.100 Safari/537.36
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8
Referer: http://localhost:8000/editor/svg-editor-es.html
Accept-Encoding: gzip, deflate, br
Accept-Language: en-US,en;q=0.9
```

Response:

```
HTTP/1.1 500 Internal Server Error
Date: Wed, 26 Sep 2018 03:34:32 GMT
Server: Apache/2.2.22
Vary: Accept-Encoding
Content-Encoding: gzip
Content-Length: 339
Connection: close
Content-Type: text/html; charset=iso-8859-1
```

**Comment by adrianbj** [comment]
Sorry, that should be fixed.

I am tempted to give up soon - I don't want to keep wasting your time with it - it's hard to test when you can't reproduce locally :)

**Comment by brettz9** [comment]
Don't worry on my account, but I know it's not easy when one is not able to reproduce... And maybe it is peculiar to my location/VPN....

**Comment by brettz9** [comment]
Ok, I'm going to close then, as it appears to be working as reasonably possible, and I've done what I can on Openclipart, though disabling it for now (and will hopefully hear back about their now missing CORS headers which their API indicated they supported but currently does not).

**Comment by brettz9** [comment]
And thanks again for your work on reporting these issues and your work on the IAN site!

**Comment by rejon** [comment]
will get added back to @openclipart
